tools/run_tests/xds_k8s_test_driver/framework/test_app/base_runner.py

```diff
@@ -35,14 +46,109 @@ class KubernetesBaseRunner:
     def __init__(self,
                  k8s_namespace,
                  namespace_template=None,
-                 reuse_namespace=False):
+                 reuse_namespace=False,
+                 case_name=None,
+                 pod_label_selector=None):
         # Kubernetes namespaced resources manager
         self.k8s_namespace: k8s.KubernetesNamespace = k8s_namespace
         self.reuse_namespace = reuse_namespace
         self.namespace_template = namespace_template or 'namespace.yaml'
+        self.case_name: str = case_name or ""
+        self.pod_label_selector: str = pod_label_selector or ""
 
         # Mutable state
         self.namespace: Optional[k8s.V1Namespace] = None
+        self._log_stop_event: threading.Event = threading.Event()
+
+    def _await_container_ready(self, core: client.CoreV1Api, pod_name: str,
+                               container_name: str, namespace: str) -> None:
+        while not self._log_stop_event.is_set():
+            try:
+                pod_status = core.read_namespaced_pod_status(
+                    pod_name, namespace).status
+                container = [
+                    container for container in pod_status.container_statuses
+                    if container.name == container_name
+                ][0]
+                if not container.ready:
+                    time.sleep(_BACKOFF_SECONDS)
+                else:
+                    break
+            except kubernetes.client.rest.ApiException:
+                time.sleep(_BACKOFF_SECONDS)
+
+    def _log_container(self, core: client.CoreV1Api, pod_name: str,
+                       container_name: str, namespace: str) -> None:
+        restarted = False
+        logfile = os.path.join(
+            _TEST_LOG_BASE_DIR,
+            f"{self.case_name}.{pod_name}.{container_name}.sponge_log.txt")
+
+        self._await_container_ready(core, pod_name, container_name, namespace)
+
+        with open(logfile, "w") as f:
+            while not self._log_stop_event.is_set():
+                try:
+                    if restarted:
+                        f.write(
+                            "Restarted log fetching. Attempting to read from the beginning, but truncation may have occurred.\n"
+                        )
```

Comment:
Just curious, what could cause the pod to crash?

Reply:
This isn't actually about the pod crashing. This is about the log query to the `kube-api-server` instance failing. Anecdotally, I've seen this happen pretty frequently in GKE when a container is creating lots of logs. The query will just repeatedly time out.

I've renamed this variable to `query_restarted` to better indicate what it is supposed to be tracking.

Reply:
Sounds good.